`unused_async_trait_impl` is not working well with `!`

### Summary

`unused_async_trait_impl` should probably recognize `!` type and not trigger on it

### Lint Name

unused_async_trait_impl

### Reproducer

I tried this code:

```rust
impl ReadAtAsync for ! {
    async fn read_at<B>(&self, _buf: B, _offset: u64) -> io::Result<B>
    where
        AsyncReadBytes<B>: From<B>,
        B: AsMut<[u8]> + Unpin + 'static,
    {
        unreachable!("Is never called")
    }
}
```

I saw this happen:

```
warning: unused `async` for async trait impl function with no `.await` statements
   --> crates/farmer/ab-farmer-components/src/lib.rs:150:5
    |
150 | /     async fn read_at<B>(&self, _buf: B, _offset: u64) -> io::Result<B>
151 | |     where
152 | |         AsyncReadBytes<B>: From<B>,
153 | |         B: AsMut<[u8]> + Unpin + 'static,
154 | |     {
155 | |         unreachable!("Is never called")
156 | |     }
    | |_____^
    |
    = note: `std::future::ready` creates a `Future` which returns the value immediately when `poll`ed
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_async_trait_impl
    = note: `-W clippy::unused-async-trait-impl` implied by `-W clippy::pedantic`
    = help: to override `-W clippy::pedantic` add `#[allow(clippy::unused_async_trait_impl)]`
help: consider removing the `async` from this function and returning `impl Future<Output = io::Result<B>>` instead
    |
150 ~     fn read_at<B>(&self, _buf: B, _offset: u64) -> impl Future<Output = io::Result<B>>
151 |     where
...
154 |     {
155 ~         std::future::ready(unreachable!("Is never called"))
    |
```

Applying the suggestion leads to a different warning though:
```
   --> crates/farmer/ab-farmer-components/src/lib.rs:155:9
    |
155 |         ready(unreachable!("Is never called"))
    |         ^^^^^ ------------------------------- any code following this expression is unreachable
    |         |
    |         unreachable call
    |
    = note: `#[warn(unreachable_code)]` (part of `#[warn(unused)]`) on by default
```

I think it'd be better to special-case `!` and not trigger this lint on it at all.

### Version

```text
rustc 1.98.0-nightly (e7815e522 2026-06-04)
binary: rustc
commit-hash: e7815e522ecc746592fee32f50478f521333b503
commit-date: 2026-06-04
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.6
```

### Additional Labels

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

> So maybe the lint should still trigger, but suggest just `unreachable!()` instead of `ready(unreachable!())`?

That will not compile at all in today's Rust:
```
error[E0277]: `!` is not a future
   --> crates/farmer/ab-farmer-components/src/lib.rs:154:52
    |
154 |     fn read_at<B>(&self, _buf: B, _offset: u64) -> impl Future<Output = io::Result<B>>
    |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `!` is not a future
...
159 |         unreachable!("Is never called")
    |         ------------------------------- return type was inferred to be `!` here
    |
    = help: the trait `futures::Future` is not implemented for `!`
```

--- a/clippy_lints/src/unused_async.rs
+++ b/clippy_lints/src/unused_async.rs
@@ -285,6 +285,20 @@ impl<'tcx> LateLintPass<'tcx> for UnusedAsync {
                 && let ExprKind::Block(block, _) = inner.kind
                 && let Some(tail_expr) = block.expr
             {
+                // check the impl method self type/item
+                if let Ok(args) = cx.tcx.try_normalize_erasing_regions(
+                    cx.typing_env(),
+                    cx.tcx
+                        .fn_sig(impl_item.owner_id.def_id)
+                        .instantiate_identity()
+                        .map(|x| cx.tcx.instantiate_bound_regions_with_erased(x.inputs_and_output())),
+                ) && args[..args.len() - 1].iter().any(|&x| {
+                    !x.peel_refs()
+                        .is_inhabited_from(cx.tcx, impl_item.owner_id.def_id.to_def_id(), cx.typing_env())
+                }) {
+                    return;
+                }
+
                 span_lint_and_then(
                     cx,
                     UNUSED_ASYNC_TRAIT_IMPL,

--- a/clippy_lints/src/unused_async.rs
+++ b/clippy_lints/src/unused_async.rs
@@ -285,6 +285,23 @@ impl<'tcx> LateLintPass<'tcx> for UnusedAsync {
                 && let ExprKind::Block(block, _) = inner.kind
                 && let Some(tail_expr) = block.expr
             {
+                // check the impl method self type/item
+                if let Ok(args) = cx.tcx.try_normalize_erasing_regions(
+                    cx.typing_env(),
+                    cx.tcx
+                        .fn_sig(impl_item.owner_id.def_id)
+                        .instantiate_identity()
+                        .map(|x| cx.tcx.instantiate_bound_regions_with_erased(x.inputs_and_output())),
+                ) && args[..args.len() - 1].iter().any(|&x| {
+                    !x.peel_refs().is_inhabited_from(
+                        cx.tcx,
+                        cx.tcx.parent_module_from_def_id(impl_item.owner_id.def_id),
+                        cx.typing_env(),
+                    )
+                }) {
+                    return;
+                }
+
                 span_lint_and_then(
                     cx,
                     UNUSED_ASYNC_TRAIT_IMPL,

--- a/tests/ui/unused_async_trait_impl.fixed
+++ b/tests/ui/unused_async_trait_impl.fixed
@@ -1,4 +1,5 @@
 #![warn(clippy::unused_async_trait_impl)]
+#![feature(never_type)]
 
 trait HasAsyncMethod {
     async fn do_something() -> u32;
@@ -115,7 +116,6 @@ mod issue17179 {
         fn do_something() -> impl Future<Output = u32> {
             //~^ unused_async_trait_impl
 
-            // Test that local functions are not touched by the suggestion.
             fn local_func() -> u32 {
                 if 5 == 2 {
                     return 1;
@@ -123,7 +123,6 @@ mod issue17179 {
                 2
             }
 
-            // Test that we do not change the tail expr or return in a (unrelated) closure.
             let f = || {
                 if 5 == 2 {
                     return 1;
@@ -131,12 +130,32 @@ mod issue17179 {
                 2
             };
 
-            // However the following return statement and tail expression should be changed.
             if f() == 5 {
                 return std::future::ready(3);
             }
 
             std::future::ready(5)
+        }
+    }
+}
+
+mod issue17162 {
+    trait AsyncTraitWithSelf {
+        async fn do_something(&self) -> u32;
+    }
+
+    impl AsyncTraitWithSelf for ! {
+        async fn do_something(&self) -> u32 {
+            unreachable!()
+        }
+    }
+
+    // so this should lint, cause no self parameter, can be called via <! as
+    // HasAsyncMethod>::do_something()
+    impl crate::HasAsyncMethod for ! {
+        fn do_something() -> impl Future<Output = u32> {
+            //~^ unused_async_trait_impl
+            std::future::ready(1)
         }
     }
 }

--- a/tests/ui/unused_async_trait_impl.rs
+++ b/tests/ui/unused_async_trait_impl.rs
@@ -1,4 +1,5 @@
 #![warn(clippy::unused_async_trait_impl)]
+#![feature(never_type)]
 
 trait HasAsyncMethod {
     async fn do_something() -> u32;
@@ -115,7 +116,6 @@ mod issue17179 {
         async fn do_something() -> u32 {
             //~^ unused_async_trait_impl
 
-            // Test that local functions are not touched by the suggestion.
             fn local_func() -> u32 {
                 if 5 == 2 {
                     return 1;
@@ -123,7 +123,6 @@ mod issue17179 {
                 2
             }
 
-            // Test that we do not change the tail expr or return in a (unrelated) closure.
             let f = || {
                 if 5 == 2 {
                     return 1;
@@ -131,12 +130,32 @@ mod issue17179 {
                 2
             };
 
-            // However the following return statement and tail expression should be changed.
             if f() == 5 {
                 return 3;
             }
 
             5
+        }
+    }
+}
+
+mod issue17162 {
+    trait AsyncTraitWithSelf {
+        async fn do_something(&self) -> u32;
+    }
+
+    impl AsyncTraitWithSelf for ! {
+        async fn do_something(&self) -> u32 {
+            unreachable!()
+        }
+    }
+
+    // so this should lint, cause no self parameter, can be called via <! as
+    // HasAsyncMethod>::do_something()
+    impl crate::HasAsyncMethod for ! {
+        async fn do_something() -> u32 {
+            //~^ unused_async_trait_impl
+            1
         }
     }
 }

--- a/tests/ui/unused_async_trait_impl.stderr
+++ b/tests/ui/unused_async_trait_impl.stderr
@@ -1,5 +1,5 @@
 error: unused `async` for async trait impl function with no `.await` statements
-  --> tests/ui/unused_async_trait_impl.rs:16:9
+  --> tests/ui/unused_async_trait_impl.rs:17:9
    |
 LL | /         async fn do_something() -> u32 {
 LL | |
@@ -18,7 +18,7 @@ LL ~             std::future::ready(1)
    |
 
 error: unused `async` for async trait impl function with no `.await` statements
-  --> tests/ui/unused_async_trait_impl.rs:40:9
+  --> tests/ui/unused_async_trait_impl.rs:41:9
    |
 LL | /         async fn do_something() -> u32 {
 LL | |
@@ -40,7 +40,7 @@ LL ~             std::future::ready(x)
    |
 
 error: unused `async` for async trait impl function with no `.await` statements
-  --> tests/ui/unused_async_trait_impl.rs:63:9
+  --> tests/ui/unused_async_trait_impl.rs:64:9
    |
 LL | /         async fn do_something() -> u32 {
 LL | |
@@ -57,7 +57,7 @@ LL ~             std::future::ready(5)
    |
 
 error: unused `async` for async trait impl function with no `.await` statements
-  --> tests/ui/unused_async_trait_impl.rs:97:9
+  --> tests/ui/unused_async_trait_impl.rs:98:9
    |
 LL | /         async fn do_something() -> vec_ty!(u32) {
 LL | |
@@ -74,7 +74,7 @@ LL ~             std::future::ready(Vec::new())
    |
 
 error: unused `async` for async trait impl function with no `.await` statements
-  --> tests/ui/unused_async_trait_impl.rs:104:9
+  --> tests/ui/unused_async_trait_impl.rs:105:9
    |
 LL | /         async fn do_something() -> Vec<u32> {
 LL | |
@@ -91,9 +91,12 @@ LL ~             std::future::ready(vec![])
    |
 
 error: unused `async` for async trait impl function with no `.await` statements
-  --> tests/ui/unused_async_trait_impl.rs:115:9
+  --> tests/ui/unused_async_trait_impl.rs:116:9
    |
 LL | /         async fn do_something() -> u32 {
+LL | |
+LL | |
+LL | |             fn local_func() -> u32 {
 ...  |
 LL | |             5
 LL | |         }
@@ -112,5 +115,22 @@ LL |
 LL ~             std::future::ready(5)
    |
 
-error: aborting due to 6 previous errors
+error: unused `async` for async trait impl function with no `.await` statements
+  --> tests/ui/unused_async_trait_impl.rs:156:9
+   |
+LL | /         async fn do_something() -> u32 {
+LL | |
+LL | |             1
+LL | |         }
+   | |_________^
+   |
+   = note: `std::future::ready` creates a `Future` which returns the value immediately when `poll`ed
+help: consider removing the `async` from this function and returning `impl Future<Output = u32>` instead
+   |
+LL ~         fn do_something() -> impl Future<Output = u32> {
+LL |
+LL ~             std::future::ready(1)
+   |
+
+error: aborting due to 7 previous errors
 

